File: ios/Classes/FakeAnalyticsPlugin.m
#import "FakeAnalyticsPlugin.h"
#import <FakeTalkingdataAnalytics/TalkingData.h>

@implementation FakeAnalyticsPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"v7lin.github.io/fake_analytics"
                                     binaryMessenger:[registrar messenger]];
    FakeAnalyticsPlugin* instance = [[FakeAnalyticsPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

static NSString * const METHOD_STARTWORK = @"startWork";
static NSString * const METHOD_SIGNUP = @"signUp";
static NSString * const METHOD_SIGNIN = @"signIn";
static NSString * const METHOD_TRACKEVENT = @"trackEvent";
static NSString * const METHOD_TRACKPLACEORDER = @"trackPlaceOrder";
static NSString * const METHOD_TRACKPAYORDER = @"trackPayOrder";
static NSString * const METHOD_STARTPAGETRACKING = @"startPageTracking";
static NSString * const METHOD_STOPPAGETRACKING = @"stopPageTracking";

static NSString * const ARGUMENT_KEY_APPID = @"appId";
static NSString * const ARGUMENT_KEY_CHANNELID = @"channelId";
static NSString * const ARGUMENT_KEY_ENABLEDEBUG = @"enableDebug";

static NSString * const ARGUMENT_KEY_TYPE = @"type";
static NSString * const ARGUMENT_KEY_UID = @"uid";
static NSString * const ARGUMENT_KEY_NAME = @"name";
static NSString * const ARGUMENT_KEY_ORDERID = @"orderId";
static NSString * const ARGUMENT_KEY_TOTAL = @"total";
static NSString * const ARGUMENT_KEY_CURRENCYTYPE = @"currencyType";
static NSString * const ARGUMENT_KEY_PAYTYPE = @"payType";
static NSString * const ARGUMENT_KEY_EVENTID = @"eventId";
static NSString * const ARGUMENT_KEY_EVENTLABEL = @"eventLabel";
static NSString * const ARGUMENT_KEY_EVENTPARAMS = @"eventParams";

static NSString * const ARGUMENT_KEY_PAGENAME = @"pageName";

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([METHOD_STARTWORK isEqualToString:call.method]) {
        NSString * appId = call.arguments[ARGUMENT_KEY_APPID];
        NSString * channelId = call.arguments[ARGUMENT_KEY_CHANNELID];
        NSNumber * enableDebug = call.arguments[ARGUMENT_KEY_ENABLEDEBUG];
        [TalkingData sessionStarted:appId withChannelId:channelId];
        [TalkingData setExceptionReportEnabled:YES];
        [TalkingData setLogEnabled:enableDebug.boolValue];
        result(nil);
    } else if ([METHOD_SIGNUP isEqualToString:call.method]) {
        NSNumber * type = call.arguments[ARGUMENT_KEY_TYPE];
        NSString * uid = call.arguments[ARGUMENT_KEY_UID];
        NSString * name = call.arguments[ARGUMENT_KEY_NAME];
        [TalkingData onRegister:uid type:type.unsignedIntegerValue name:name];
        result(nil);
    } else if ([METHOD_SIGNIN isEqualToString:call.method]) {
        NSNumber * type = call.arguments[ARGUMENT_KEY_TYPE];
        NSString * uid = call.arguments[ARGUMENT_KEY_UID];
        NSString * name = call.arguments[ARGUMENT_KEY_NAME];
        [TalkingData onLogin:uid type:type.unsignedIntegerValue name:name];
        result(nil);
    } else if ([METHOD_TRACKEVENT isEqualToString:call.method]) {
        NSString * eventId = call.arguments[ARGUMENT_KEY_EVENTID];
        NSString * eventLabel = call.arguments[ARGUMENT_KEY_EVENTLABEL];
        NSDictionary * eventParams = call.arguments[ARGUMENT_KEY_EVENTPARAMS];
        [TalkingData trackEvent:eventId label:eventLabel parameters:eventParams];
        result(nil);
    } else if ([METHOD_TRACKPLACEORDER isEqualToString:call.method]) {
        NSString * uid = call.arguments[ARGUMENT_KEY_UID];
        NSString * orderId = call.arguments[ARGUMENT_KEY_ORDERID];
        NSNumber * total = call.arguments[ARGUMENT_KEY_TOTAL];
        NSString * currencyType = call.arguments[ARGUMENT_KEY_CURRENCYTYPE];
        TalkingDataOrder *order = [TalkingDataOrder createOrder:orderId total:[total intValue] currencyType:currencyType];
        [TalkingData onPlaceOrder:uid order:order];
    } else if ([METHOD_TRACKPAYORDER isEqualToString:call.method]) {
        NSString * uid = call.arguments[ARGUMENT_KEY_UID];
        NSString * payType = call.arguments[ARGUMENT_KEY_PAYTYPE];
        NSString * orderId = call.arguments[ARGUMENT_KEY_ORDERID];
        NSNumber * total = call.arguments[ARGUMENT_KEY_TOTAL];
        NSString * currencyType = call.arguments[ARGUMENT_KEY_CURRENCYTYPE];
        TalkingDataOrder *order = [TalkingDataOrder createOrder:orderId total:[total intValue] currencyType:currencyType];
        [TalkingData onOrderPaySucc:uid payType:payType order:order];
    } else if ([METHOD_STARTPAGETRACKING isEqualToString:call.method]) {
        NSString * pageName = call.arguments[ARGUMENT_KEY_PAGENAME];
        [TalkingData trackPageBegin:pageName];
        result(nil);
    } else if ([METHOD_STOPPAGETRACKING isEqualToString:call.method]) {
        NSString * pageName = call.arguments[ARGUMENT_KEY_PAGENAME];
        [TalkingData trackPageEnd:pageName];
        result(nil);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

@end

File: android/src/main/java/io/github/v7lin/fakeanalytics/FakeAnalyticsPlugin.java
package io.github.v7lin.fakeanalytics;

import android.text.TextUtils;

import com.tendcloud.tenddata.TCAgent;
import com.tendcloud.tenddata.TDAccount;

import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/**
 * FakeAnalyticsPlugin
 */
public class FakeAnalyticsPlugin implements MethodCallHandler {
    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "v7lin.github.io/fake_analytics");
        FakeAnalyticsPlugin plugin = new FakeAnalyticsPlugin(registrar);
        channel.setMethodCallHandler(plugin);
    }

    private static final String METHOD_STARTWORK = "startWork";
    private static final String METHOD_SIGNUP = "signUp";
    private static final String METHOD_SIGNIN = "signIn";
    private static final String METHOD_TRACKEVENT = "trackEvent";
    private static final String METHOD_STARTPAGETRACKING = "startPageTracking";
    private static final String METHOD_STOPPAGETRACKING = "stopPageTracking";

    private static final String ARGUMENT_KEY_APPID = "appId";
    private static final String ARGUMENT_KEY_CHANNELID = "channelId";
    private static final String ARGUMENT_KEY_ENABLEDEBUG = "enableDebug";

    private static final String ARGUMENT_KEY_TYPE = "type";
    private static final String ARGUMENT_KEY_UID = "uid";
    private static final String ARGUMENT_KEY_NAME = "name";

    private static final String ARGUMENT_KEY_EVENTID = "eventId";
    private static final String ARGUMENT_KEY_EVENTLABEL = "eventLabel";
    private static final String ARGUMENT_KEY_EVENTPARAMS = "eventParams";

    private static final String ARGUMENT_KEY_PAGENAME = "pageName";

    private final Registrar registrar;

    private FakeAnalyticsPlugin(Registrar registrar) {
        this.registrar = registrar;
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        if (METHOD_STARTWORK.equals(call.method)) {
            String appId = call.argument(ARGUMENT_KEY_APPID);
            String channelId = call.argument(ARGUMENT_KEY_CHANNELID);
            boolean enableDebug = call.argument(ARGUMENT_KEY_ENABLEDEBUG);
            TCAgent.init(registrar.context(), appId, channelId);
            TCAgent.setReportUncaughtExceptions(true);
            TCAgent.LOG_ON = enableDebug;
            result.success(null);
        } else if (METHOD_SIGNUP.equals(call.method)) {
            int type = call.argument(ARGUMENT_KEY_TYPE);
            String uid = call.argument(ARGUMENT_KEY_UID);
            String name = call.argument(ARGUMENT_KEY_NAME);
            TCAgent.onRegister(uid, convertToAccountType(type), name);
            result.success(null);
        } else if (METHOD_SIGNIN.equals(call.method)) {
            int type = call.argument(ARGUMENT_KEY_TYPE);
            String uid = call.argument(ARGUMENT_KEY_UID);
            String name = call.argument(ARGUMENT_KEY_NAME);
            TCAgent.onLogin(uid, convertToAccountType(type), name);
            result.success(null);
        } else if (METHOD_TRACKEVENT.equals(call.method)) {
            String eventId = call.argument(ARGUMENT_KEY_EVENTID);
            String eventLabel = call.argument(ARGUMENT_KEY_EVENTLABEL);
            Map<String, Object> eventParams = call.argument(ARGUMENT_KEY_EVENTPARAMS);
            TCAgent.onEvent(registrar.context(), eventId, !TextUtils.isEmpty(eventLabel) ? eventLabel : "", eventParams);
            result.success(null);
        } else if (METHOD_STARTPAGETRACKING.equals(call.method)) {
            String pageName = call.argument(ARGUMENT_KEY_PAGENAME);
            TCAgent.onPageStart(registrar.context(), pageName);
            result.success(null);
        } else if (METHOD_STOPPAGETRACKING.equals(call.method)) {
            String pageName = call.argument(ARGUMENT_KEY_PAGENAME);
            TCAgent.onPageEnd(registrar.context(), pageName);
            result.success(null);
        } else {
            result.notImplemented();
        }
    }

    private TDAccount.AccountType convertToAccountType(int type) {
        for (TDAccount.AccountType accountType : TDAccount.AccountType.values()) {
            if (accountType.index() == type) {
                return accountType;
            }
        }
        return TDAccount.AccountType.ANONYMOUS;
    }
}
